File: 백준/Gold/15683. 감시/감시.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.*;

public class Main {
    static BufferedReader input = new BufferedReader(new InputStreamReader(System.in));
    static StringTokenizer token;

    static int n,m, cctvNum, wallNum, min;
    static int[][] graph;
    static boolean[][] secuArea;
    static int[][] cctvInfo;
    static int[][] eachCctvCase = {{0,1,2,3},{0,1},{0,1,2,3},{0,1,2,3},{0}};
    static int[] deltaX = {0,1,0,-1};
    static int[] deltaY = {1,0,-1,0};
    
    public static void main(String[] args) throws IOException {
        token = new StringTokenizer(input.readLine());
        n = Integer.parseInt(token.nextToken());
        m = Integer.parseInt(token.nextToken());

        graph = new int[n][m];
        cctvNum = 0;
        wallNum = 0;
        cctvInfo = new int[8][3]; // 빈 공간 있을 수 있음
        for (int i = 0; i < n; i++) {
            token = new StringTokenizer(input.readLine());
            for (int j = 0; j < m; j++) {
                int num = Integer.parseInt(token.nextToken());
                graph[i][j] = num;
                if (num > 0 && num <=5){
                    cctvInfo[cctvNum] = new int[]{num,j,i};
                    cctvNum++;
                }else if(num == 6){
                    wallNum++;
                }
            }
        }

        min = Integer.MAX_VALUE;
        Perm(0,new int[cctvNum]);
        System.out.println(min);
    }
    public static void Perm(int cnt, int[] save){
        if (cnt == cctvNum){
            secuArea = new boolean[n][m];
            for (int i = 0; i < cctvNum; i++) {
                order(cctvInfo[i][0],cctvInfo[i][1],cctvInfo[i][2],save[i]);
            }

            int res = 0;
            for (int i=0; i<n; i++){
                for (int j=0; j<m; j++){
                    if (secuArea[i][j]){
                        res++;
                    }
                }
            }

            if (n*m-res-wallNum < min){
                min = n*m-res-wallNum;
            }
            return;
        }
        for (int i = 0; i < eachCctvCase[cctvInfo[cnt][0] - 1].length; i++) {
            save[cnt] = eachCctvCase[cctvInfo[cnt][0] - 1][i];
            Perm(cnt + 1, save);
        }
    }

    public static void order(int cctvId, int x, int y, int direct) {
        switch (cctvId) {
            case 1:
                move(x, y, direct);
                break;
            case 2:
                move(x, y, direct);
                move(x, y, (direct + 2) % 4);
                break;
            case 3:
                move(x, y, direct);
                move(x, y, (direct + 1) % 4);
                break;
            case 4:
                move(x, y, direct);
                move(x, y, (direct + 1) % 4);
                move(x, y, (direct + 3) % 4);
                break;
            case 5:
                move(x, y, 0);
                move(x, y, 1);
                move(x, y, 2);
                move(x, y, 3);
                break;
        }
    }

    public static void move(int x, int y, int direct) {
        secuArea[y][x] = true;
        while (isValid(x + deltaX[direct], y + deltaY[direct]) && graph[y + deltaY[direct]][x + deltaX[direct]] != 6) {
            x += deltaX[direct];
            y += deltaY[direct];
            secuArea[y][x] = true;
        }
    }

    public static boolean isValid(int x, int y){
        return x >= 0 && x < m && y >= 0 && y < n;
    }
}

File: 백준/Gold/15683. 감시/감시.py
'''
graph[dy][dx] == 6:을 == '6':이라 해서 틀리고 찾는 것도 진짜 한참 찾았다 하..
'''
import sys

def is_valid(x,y):
    return 0<=x<M and 0<=y<N

delta_x = [0,1,0,-1]
delta_y = [-1,0,1,0]

N,M = map(int, sys.stdin.readline().split())
graph = []
for _ in range(N):
    graph.append(list(map(int, sys.stdin.readline().split())))

cctv_dic = {}
cctv_dic[1] = [
    [0],
    [1],
    [2],
    [3]
]
cctv_dic[2] = [
    [1,3],
    [0,2]
]
cctv_dic[3] = [
    [0,1],
    [1,2],
    [2,3],
    [3,0]
]
cctv_dic[4] = [
    [0,2,3],
    [0,1,3],
    [0,1,2],
    [1,2,3]
]
cctv_dic[5] = [
    [0,1,2,3]
]

cctv_idxs = []
cctv_types = []
for i in range(N):
    for j in range(M):
        if 1 <= graph[i][j] <= 5:
            cctv_idxs.append([j,i])
            cctv_types.append(graph[i][j])

def back_track(n,lst):
    global min_res

    if n == len(cctv_idxs):
        visited = []
        for _ in range(N):
            visited.append([0]*M)

        for cctv_id in range(len(cctv_idxs)):
            cctv_x,cctv_y = cctv_idxs[cctv_id]
            type = cctv_types[cctv_id]
            cctv_directs = cctv_dic[type][lst[cctv_id]]

            for cctv_direct in cctv_directs:
                x,y = cctv_x,cctv_y
                while True:
                    dx = x + delta_x[cctv_direct]
                    dy = y + delta_y[cctv_direct]
                    if not is_valid(dx,dy):
                        break
                    if graph[dy][dx] == 6:
                        break
                    visited[dy][dx] = 1
                    x,y = dx,dy

        cnt = 0
        for i in range(N):
            for j in range(M):
                if not visited[i][j] and graph[i][j]==0:
                    cnt += 1

        min_res = min(min_res, cnt)
        return

    for i in range(len(cctv_dic[cctv_types[n]])):
        back_track(n+1,lst+[i])


min_res = N*M
back_track(0,[])
print(min_res)